R/data-combined.R

```diff
@@ -0,0 +1,680 @@
+
+    .extractNames = function(data = NULL) {
+      # return early if there is no data
+      if (is.null(data)) {
+        return(NULL)
+      }
+
+      data %>%
+        dplyr::pull(name) %>%
+        unique() %>%
+        sort()
+    },
+
+    # Keep internal copies of the raw values before changing them using
+    # transformations with given offsets and scale factors.
+    #
+    # This way, when transformation parameter values are updated, the raw data
+    # is still there to be transformed with these new parameters.
+
+    .extractXYData = function(data = NULL) {
+      # return early if there is no data
+      if (is.null(data)) {
+        return(NULL)
+      }
+
+      data %>%
+        dplyr::mutate(
+          xOriginalValues      = xValues,
+          yOriginalValues      = yValues,
+          yOriginalErrorValues = yErrorValues
+        )
+    },
+
+    # Extract offsets and scale factors used while data transformations
+    #
+    # Since the user might have entered distinct transformation parameters for
+    # each dataset, the combined dataframe needs to be grouped by dataset names.
+
+    .extractTransforms = function(data = NULL) {
+      # return early if there is no data
+      if (is.null(data)) {
+        return(NULL)
+      }
+
+      # select the unique identifier and transformation parameter columns and then
+      # group the combined dataframe by unique datasets and then
+      # within each dataset retain distinct row combinations and then
+      # ungroup the dataframe
+      data %>%
+        dplyr::select(name, dplyr::matches("offset|scale")) %>%
+        dplyr::group_by(name) %>%
+        dplyr::distinct() %>%
+        dplyr::ungroup()
+    },
+
+    # clean dataframe before returning it to the user
+
+    .cleanDF = function(data = NULL) {
+      # return early if there is no data
+      if (is.null(data)) {
+        return(NULL)
+      }
+
+      # having a consistent column order
+      data <- data %>%
+        dplyr::select(
+          # all identifying columns
+          name,
+          group,
+          dataType,
+          # everything related to the X-variable
+          "xValues", "xUnit", "xDimension", dplyr::matches("^x"),
+          # everything related to the Y-variable
+          "yValues", "yUnit", "yDimension", dplyr::matches("^y"),
+          # all other columns go after that (meta data, etc.)
+          dplyr::everything()
+        ) %>%
+        # the following columns are no longer necessary
+        #
+        # retaining the offset and scale factor parameters might confuse the
+        # user about whether the transformations are supposed to be carried out
+        # by the user using these values or these transformations have already
+        # been carried out
+        #
+        # internal copies of raw data and redundant paths columns also need to
+        # be left out
+        dplyr::select(
+          -dplyr::matches("^paths$"),
```

Comment:
Can you explain the path and original? I am assuming they are added somewhere programatically but I can't find it
also what happens  if some meta data have the name original in them? they are going to be removed. It is too much of a common name to remove it like this. Maybe our internal keyword should be encapsulated in marker (typically succession of one or two chars such as &%original&%. And you define constants for these keywords  so that you do not have to remembner what this is